File: backend/app/schemas/mail.py
"""
Mail System Schemas — Pydantic models for request/response validation
"""
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, EmailStr


# ─────────────────────────────────────────────
# Email Account
# ─────────────────────────────────────────────

class EmailAccountCreate(BaseModel):
    display_name: str
    email_address: EmailStr
    smtp_host: str
    smtp_port: int = 587
    smtp_use_tls: bool = True
    imap_host: str
    imap_port: int = 993
    imap_use_ssl: bool = True
    username: str
    password: str  # plain — will be encrypted before storage


class EmailAccountUpdate(BaseModel):
    display_name: Optional[str] = None
    smtp_host: Optional[str] = None
    smtp_port: Optional[int] = None
    smtp_use_tls: Optional[bool] = None
    imap_host: Optional[str] = None
    imap_port: Optional[int] = None
    imap_use_ssl: Optional[bool] = None
    username: Optional[str] = None
    password: Optional[str] = None  # plain — will be encrypted before storage


class EmailAccountResponse(BaseModel):
    id: int
    user_id: int
    display_name: str
    email_address: str
    smtp_host: str
    smtp_port: int
    smtp_use_tls: bool
    imap_host: str
    imap_port: int
    imap_use_ssl: bool
    username: str
    is_active: bool
    is_verified: bool
    last_sync_at: Optional[datetime] = None
    created_at: datetime

    model_config = {"from_attributes": True}


# ─────────────────────────────────────────────
# Attachment
# ─────────────────────────────────────────────

class AttachmentResponse(BaseModel):
    id: int
    email_id: int
    filename: str
    content_type: Optional[str] = None
    size_bytes: Optional[int] = None

    model_config = {"from_attributes": True}


# ─────────────────────────────────────────────
# Email
# ─────────────────────────────────────────────

class EmailSend(BaseModel):
    account_id: int
    to_addresses: List[EmailStr]
    cc_addresses: Optional[List[EmailStr]] = []
    bcc_addresses: Optional[List[EmailStr]] = []
    subject: str
    body_html: str
    body_text: Optional[str] = None
    reply_to_email_id: Optional[int] = None  # for threading
    tenant_id: Optional[int] = None


class DraftSave(BaseModel):
    account_id: int
    to_addresses: Optional[List[str]] = []
    cc_addresses: Optional[List[str]] = []
    bcc_addresses: Optional[List[str]] = []
    subject: Optional[str] = ""
    body_html: Optional[str] = ""
    body_text: Optional[str] = None
    draft_id: Optional[int] = None  # if updating existing draft


class EmailResponse(BaseModel):
    id: int
    account_id: int
    thread_id: Optional[int] = None
    message_id: Optional[str] = None
    folder: str
    from_address: str
    from_name: Optional[str] = None
    to_addresses: str   # JSON string
    cc_addresses: Optional[str] = None
    bcc_addresses: Optional[str] = None
    subject: str
    body_text: Optional[str] = None
    body_html: Optional[str] = None
    is_read: bool
    is_starred: bool
    is_draft: bool
    date: datetime
    created_at: datetime
    attachments: List[AttachmentResponse] = []

    model_config = {"from_attributes": True}


class EmailListItem(BaseModel):
    """Lightweight email for list view."""
    id: int
    account_id: int
    thread_id: Optional[int] = None
    folder: str
    from_address: str
    from_name: Optional[str] = None
    to_addresses: str
    subject: str
    body_text: Optional[str] = None
    is_read: bool
    is_starred: bool
    is_draft: bool
    date: datetime
    attachment_count: int = 0

    model_config = {"from_attributes": True}


# ─────────────────────────────────────────────
# Thread
# ─────────────────────────────────────────────

class ThreadResponse(BaseModel):
    id: int
    account_id: int
    subject: str
    last_message_at: datetime
    message_count: int
    has_unread: bool
    tenant_id: Optional[int] = None

    model_config = {"from_attributes": True}


# ─────────────────────────────────────────────
# Misc
# ─────────────────────────────────────────────

class MailboxStats(BaseModel):
    inbox_unread: int
    drafts_count: int
    sent_count: int
    trash_count: int


class ConnectionTestResult(BaseModel):
    smtp_ok: bool
    imap_ok: bool
    smtp_error: Optional[str] = None
    imap_error: Optional[str] = None


class MarkReadPayload(BaseModel):
    email_ids: List[int]
    is_read: bool = True


class MoveToTrashPayload(BaseModel):
    email_ids: List[int]

File: frontend/src/store/mail.ts
import { create } from "zustand";
import { api } from "../lib/api";

export type EmailAccount = {
  id: number;
  user_id: number;
  display_name: string;
  email_address: string;
  smtp_host: string;
  smtp_port: number;
  smtp_use_tls: boolean;
  imap_host: string;
  imap_port: number;
  imap_use_ssl: boolean;
  username: string;
  is_active: boolean;
  is_verified: boolean;
  last_sync_at: string | null;
  created_at: string;
};

export type EmailListItem = {
  id: number;
  account_id: number;
  thread_id: number | null;
  folder: string;
  from_address: string;
  from_name: string | null;
  to_addresses: string; // JSON string
  subject: string;
  body_text: string | null;
  is_read: boolean;
  is_starred: boolean;
  is_draft: boolean;
  date: string;
  attachment_count: number;
};

export type EmailDetail = EmailListItem & {
  cc_addresses: string | null;
  bcc_addresses: string | null;
  body_html: string | null;
  attachments: { id: number; filename: string; content_type: string | null; size_bytes: number | null }[];
};

export type MailboxStats = {
  inbox_unread: number;
  drafts_count: number;
  sent_count: number;
  trash_count: number;
};

type MailState = {
  accounts: EmailAccount[];
  activeAccountId: number | null;
  activeFolder: string;
  emails: EmailListItem[];
  selectedEmail: EmailDetail | null;
  stats: MailboxStats | null;
  loading: boolean;
  syncing: boolean;

  fetchAccounts: () => Promise<void>;
  setActiveAccount: (id: number) => void;
  setActiveFolder: (folder: string) => void;
  fetchEmails: (accountId: number, folder: string, search?: string) => Promise<void>;
  fetchEmail: (accountId: number, emailId: number) => Promise<void>;
  fetchStats: (accountId: number) => Promise<void>;
  syncInbox: (accountId: number) => Promise<number>;
  markRead: (accountId: number, emailIds: number[], isRead: boolean) => Promise<void>;
  moveToTrash: (accountId: number, emailIds: number[]) => Promise<void>;
  deleteEmail: (accountId: number, emailId: number) => Promise<void>;
  clearSelectedEmail: () => void;
};

export const useMailStore = create<MailState>((set, get) => ({
  accounts: [],
  activeAccountId: null,
  activeFolder: "inbox",
  emails: [],
  selectedEmail: null,
  stats: null,
  loading: false,
  syncing: false,

  fetchAccounts: async () => {
    try {
      const { data } = await api.get("/mail/accounts");
      const list: EmailAccount[] = Array.isArray(data) ? data : [];
      const currentId = get().activeAccountId;

      // If the currently active account no longer exists in the refreshed list,
      // reset to the first available account (or null if none).
      const idStillValid = currentId !== null && list.some((a) => a.id === currentId);
      const nextActiveId = idStillValid
        ? currentId
        : list.length > 0
        ? list[0].id
        : null;

      set({
        accounts: list,
        activeAccountId: nextActiveId,
        // Clear stale email/stats data if the active account changed
        ...(nextActiveId !== currentId
          ? { emails: [], selectedEmail: null, stats: null }
          : {}),
      });
    } catch (err: unknown) {
      // Silently keep accounts as empty array — 401 means not yet authed,
      // 500 means backend issue; both are handled gracefully in the UI.
      const status = (err as { response?: { status?: number } })?.response?.status;
      if (status !== 401 && status !== 500) {
        console.error("[mail] fetchAccounts failed:", err);
      }
      set({ accounts: [] });
    }
  },

  setActiveAccount: (id) => {
    set({ activeAccountId: id, selectedEmail: null, emails: [] });
  },

  setActiveFolder: (folder) => {
    set({ activeFolder: folder, selectedEmail: null, emails: [] });
  },

  fetchEmails: async (accountId, folder, search) => {
    set({ loading: true });
    try {
      const params: Record<string, string> = { folder };
      if (search) params.search = search;
      const { data } = await api.get(`/mail/accounts/${accountId}/emails`, { params });
      set({ emails: data });
    } finally {
      set({ loading: false });
    }
  },

  fetchEmail: async (accountId, emailId) => {
    const { data } = await api.get(`/mail/accounts/${accountId}/emails/${emailId}`);
    set({ selectedEmail: data });
    // Update is_read in list
    set((state) => ({
      emails: state.emails.map((e) => (e.id === emailId ? { ...e, is_read: true } : e)),
    }));
  },

  fetchStats: async (accountId) => {
    const { data } = await api.get(`/mail/accounts/${accountId}/stats`);
    set({ stats: data });
  },

  syncInbox: async (accountId) => {
    set({ syncing: true });
    try {
      const { data } = await api.post(`/mail/accounts/${accountId}/sync`);
      return data.synced as number;
    } finally {
      set({ syncing: false });
    }
  },

  markRead: async (accountId, emailIds, isRead) => {
    await api.post(`/mail/accounts/${accountId}/mark-read`, { email_ids: emailIds, is_read: isRead });
    set((state) => ({
      emails: state.emails.map((e) =>
        emailIds.includes(e.id) ? { ...e, is_read: isRead } : e
      ),
    }));
  },

  moveToTrash: async (accountId, emailIds) => {
    await api.post(`/mail/accounts/${accountId}/trash`, { email_ids: emailIds });
    set((state) => ({
      emails: state.emails.filter((e) => !emailIds.includes(e.id)),
      selectedEmail:
        state.selectedEmail && emailIds.includes(state.selectedEmail.id)
          ? null
          : state.selectedEmail,
    }));
  },

  deleteEmail: async (accountId, emailId) => {
    await api.delete(`/mail/accounts/${accountId}/emails/${emailId}`);
    set((state) => ({
      emails: state.emails.filter((e) => e.id !== emailId),
      selectedEmail: state.selectedEmail?.id === emailId ? null : state.selectedEmail,
    }));
  },

  clearSelectedEmail: () => set({ selectedEmail: null }),
}));
